blog/terraform-post.md

```diff
@@ -0,0 +1,67 @@
+
+![][image1]
+
+# What is Terraform?
+
+Terraform is a powerful Infrastructure-as-Code (IaC) tool that allows you to define and manage your infrastructure using code. Infrastructure-as-Code, in general, is the practice of managing and provisioning infrastructure through machine-readable definition files rather than through manual processes. This means you can automate the creation, modification, and deletion of resources across various cloud providers and on-premises environments, ensuring consistency, repeatability, and version control. 
+
+Cloud companies often build out Terraform providers, which are responsible for establishing differences between the cloud/user intent, determining which API requests are necessary, making those requests, and handling any follow-up actions. 
+
+# Why do we need the AEPs for this?
+
+Let's take a look at a pseudocode version of the Terraform lifecycle.
+
+```
+userIntention := GetTerraformPlan()
+currentState := ReadResource(userIntention)
+
+if currentState != null:
+	if user wishes to delete:
+		 DeleteResource(currentState)
+	else if currentState != userIntention:
+		 UpdateResource(userIntention)
+else:
+	CreateResource(userIntention)
+```
+
+Two different sets of problems emerge and the AEPs are well-suited for solving both of them.
+
+The first set of problems revolves around the grouping and calling of APIs. There's a variety of tools in the market that will call an API given a set of parameters and an OpenAPI description (or even just a URL).Terraform doesn't expose individual APIs, but instead exposes a resource schema. This raises a variety of questions when developing a Terraform provider:
+
+* What are my resources?  
+* How do I know which APIs to call for a given resource?  
+* Is there 1 API that can be called for each CRUD action, or will I need to chain together API calls to achieve my result?  
+* How do I derive my resource schema from 4 distinct CRUD APIs? Is it simply the combination of all the body parameters from the 4 APIs?
+
+AEP APIs are resource-oriented by default, so just following the AEPs can address all of the problems above. The Terraform provider can make assumptions about which APIs should be called and how they relate to each other.
+
+# What about API responses?
+
+The second set of problems comes from reading API responses. In the pseudocode above, the Read response needs to be parsed and checked against the Terraform plan. This means that the Terraform provider needs to understand the relationship between the API response (beyond simple error checking) and the user's intention. More questions emerge:
+
+* Does the response of the Read API match the resource schema? What changes need to be made?  
+* Are there multiple ways to express a value in the cloud? If so, how can I be sure that two semantically equivalent values do not result in a diff against the Terraform plan?  
```

Comment:
What does "in the cloud" mean here?

I understand the second sentence here, just not sure what the first one is adding.  Maybe consider dropping it?